### DISCUSSION TITLE
Additional memory used by profiler

### DISCUSSION THREAD

**Discussion by toktarev** [async-profiler/async-profiler#499] [Q&A] [answered]
Hello, sorry for the new discussion (I realize that wrote about 7 during the last 3 days).

But the project is interesting and I have a lot of questions (and maybe dummy suggestions).

Is there any description of additional memory used by the profiler internally?

I saw a video presentation and as I understand Async Profiler can be used in production.

It is interesting about memory overhead during the long profiling process.

Thanks.

**Comment by apangin** [comment, marked as answer]
Depends on the application: the number of threads it has, number of classes, methods, the depth of stack traces, the intensity of profiling events, etc. The memory usage can be anywhere from a few megabytes to a few gigabytes; typically somewhere in between.

**Reply by Jongy** [reply]
@apangin , I know that currently it's not something that async-profiler does, but would you accept a PR that adds such "self-accounting" capability to async-profiler such that it's able to tell you the rough amount of memory it is consuming from the process?

**Reply by apangin** [reply]
@Jongy I don't mind the feature per se. The question is more about the specific implementation - whether it will be lightweight, localized, and wholesome. Before implementing, I'd suggest to discuss what and how it will count, and how end users will interact with it, and, more importantly, what problem it will solve.

**Reply by toktarev** [reply]
--what problem it will solve
Investigation problem, to analyze what happens internally with system (this is main profiler's goal).

**Reply by Jongy** [reply]
@apangin gotcha, I will have this in the back of my mind. Will open an issue with a proposal should I reach some mature thoughts.

**Reply by toktarev** [reply]
I'd suggest invoking all mmaps and mallocs using std::allocator (Tracking allocator).

That would let to monitor memory in one place.

Many malloc/free mmap, munmap over that code makes this memory monitoring hard.

At least this is how we monitor memory in our internal C++ products.

**Comment by toktarev** [comment]
Thanks.

Is there a possibility to ask the profiler about memory it uses using some API call?

We would add profiler's overhead monitoring to our Graphite system.

**Reply by apangin** [reply]
There is no such feature now.

**Reply by toktarev** [reply]
ok, Thanks.

This is not my personal suggestion for the suggestion.

We have other native libraries linked via JNI (native caches for example).
We ask them about their memory consumption and send reports to graphite.
This is how we can control memory using a monitoring panel.

Of course, we can just dump RSS but would be nice to have per-component consumption.

Those native caches were implemented inside the company and they track their memory consumption using C++ allocators.

Do you consider this option as something useful for the product?

**Reply by apangin** [reply]
> Do you consider this option as something useful for the product?

Probably yes, if someone would like to sponsor development of this feature.

**Reply by toktarev** [reply]
I can't be a sponsor.

But I can only start working on the corresponding PR (frankly I have some years of experience with C++) but not sure that you will approve my changes.

**Comment by toktarev** [comment]
BTW:

I see that JVM receives SIGTERM at the moment when I call "execute('flamegraph')".

Looks it is killed by OOM killer, because I don't see exception and result file is empty.

**Reply by apangin** [reply]
It's not the Linux OOM killer. The latter kills a process immediately with `SIGKILL`.

**Reply by toktarev** [reply]
http://lkml.iu.edu/hypermail/linux/kernel/0508.1/0390.html
I found it here.
And it happens only after "execute('flamegraph')".  execution call on all 750 JVMS during 1 hour of profiling.

**Reply by apangin** [reply]
What profiling interval do you set?

**Reply by toktarev** [reply]
5_000_000

**Reply by apangin** [reply]
Oh, it's clearly the misuse of the profiler. In this configuration, an application consuming 4 cpu cores on average will produce 4 * 200 * 3600, or nearly 3 million samples in one hour. It makes no sense to collect such profile as a flame graph. Even if it was enough memory to dump the profile, no browser would be able to open it.

A Flame Graph represents the aggregated picture over the entire profiling session, it does not save timestamps. So it is rarely useful for a long period. I guess that's not what you wanted. To find hotspots, a short profiling session (30-60 sec) is usually enough. To record historical data, use `jfr` output format.

**Reply by toktarev** [reply]
---no browser would be able to open it

My browser opened our flame graph.

Actually, we have no such big number of distinct stack traces, so it was not a problem for our case (except SEGTERM).

BTW: SIGTERM happened not in all 100% of cases.

**Comment by toktarev** [comment]
A couple of coins:

From my point of view, Profiler should use only some limited (pre-defined) memory.
Once the limit is achieved profiler should dump its state to the disk.

Finally, the profiler should collect all results from the disk and pass them to the user either via memory or passing path on the disk.

It would make Async Profiler considerably more resilient for long profiling process (1 hour and loner ) for heavy-leaded 
systems.

**Reply by apangin** [reply]
Profiler did exactly what you asked it to do. You asked it to collect millions of samples and aggregate them in one picture.
What you describe here is completely different mode. Async-profiler also has it. Try `jfr` format.

**Reply by toktarev** [reply]
I described some low-level architecture regardless of the configuration mode.

If it was implemented only for jfr - ok.

**Reply by apangin** [reply]
async-profiler bundle includes `jfr2flame` converter.

**Reply by toktarev** [reply]
Thanks for your assistance.

**Comment by toktarev** [comment]
Next experiment

4 nodes, 10 min of execution

3 nodes dumped flamegraph
1 node died with SIGTERM

flamegraph output size 10MB

**Comment by toktarev** [comment]
From what I see, it looks like  profiler consumes significant memory during the "dump" command execution 
(please correct me if I am wrong).

Would be nice to ask the profiler "which memory do you need for a 'dump'".

Shortly: would be nice to avoid SIGTERMS, SIGKILLS, and other killers.

**Reply by apangin** [reply]
Instead of speculating, you could just check the process `VmHWM` before dumping and after. It's a matter of few seconds, but gives much more accurate measure of memory consumption than any fancy technique of counting own mallocs and mmaps.

**Reply by toktarev** [reply]
I can't check "after".
"after" JVM died.

**Reply by krzysztofslusarski** [reply]
What output format do you use? How often do you dump the results? What is the profiling interval? 

I use async-profiler 24/7 in over 30 JVMs on production with a heavy load. I dump results every minute in the JFR format and I have no such issues as you have.

**Reply by toktarev** [reply]
I used "flamegraph".

Andrei has already explained that this is the wrong format, so I am going to switch on "jfr".

I use the discussion format of the Open Source project to suggest something
which might be useful from my point of view.

This is no speculation as Andrei pointed.

**Reply by toktarev** [reply]
A bit off-top (but anyway)
@krzysztofslusarski 

This is a very similar issue in the well-known project (RocksDB).
https://github.com/facebook/rocksdb/issues/4112

I posted it 3 years ago and many people blamed me that I can't configure RocksDB.
(it works in their productions and so on)

But RocksDB team did not close that issue for 3 years.
Finally, RocksDB's programmer posted today
"We are working on limiting memory usage and updates on that will be in".

And RocksDB is product which is used all over the world in many productions
including FaceBook.

**Reply by apangin** [reply]
By "speculation" here I meant conclusions based on guesses and assumptions rather than facts.

You say that OOM killer kills JVM, because async-profiler consumes too much memory during dumping, but haven't provided any evidence of that. As I've told, Linux OOM killer sends `SIGKILL` and logs the OOM reason to the kernel ring buffer. Have you checked `dmesg`? Since you observe `SIGTERM` instead of `SIGKILL`, I think it's not Linux OOM killer, but more likely k8s or whatever other container management system you use. In this case, check the logs of your management system.

Another evidence you could provide is the values of RSS/HWM before, during, and after dumping the profile. You clearly *can* do this in multiple ways, but for some reason prefer to argue rather than cooperate. After all, we are trying to help you, but this is hard if not impossible without research from your side. BTW, when the JVM dies, a self-accounting mechanism can't obviously help either.

**Reply by toktarev** [reply]
---we are trying to help you

Andrei, I apologize if I was incorrect.
You've already helped me yesterday by giving me advice about "JFR", thank you.

Regarding other aspects: I don't really have time for the analysis of the async profiler behaviour.
I have a lot of other work and my time also has it's value.

My posts here - just some thoughts of how we could make product more convenient.
If people don't agree with my ideas - ok.

**Comment by toktarev** [comment]
Ok, the additional investigation showed that JVM was killed by YARN NodeManager.

From what I understand YARN could do it in 2 cases:

The container (JVM) consumes too much memory.
The container (JVM) hungs in the shutting down phase (means that Async Profiler collects dump too long time for YARN).

I was not able to find somewhere in logs the actual reason why NodeManager sent SIGTERM to the container.

Anyway, in all cases, SIGTERM happens exactly during execute("asyncprofiler") execution.

This might be a problem of our cluster.

**Comment by toktarev** [comment]
Switched to JFR, but see the same problem.
Yarn killed node.

I will send detailed investigation result a little bit later.
